compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -2185,7 +2196,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     // the bad interactions of the given hack detailed in (note_1).
                     debug!("check_pat_ref: expected={:?}", expected);
                     match *expected.kind() {
-                        ty::Ref(_, r_ty, r_mutbl) if r_mutbl == mutbl => (expected, r_ty, pat_info),
+                        ty::Ref(_, r_ty, r_mutbl) if r_mutbl == mutbl => {
+                            let pat_info = if r_mutbl == Mutability::Not
+                                && ((pat.span.at_least_rust_2024()
+                                    && self.tcx.features().ref_pat_eat_one_layer_2024)
+                                    || self.tcx.features().ref_pat_everywhere)
+                            {
+                                PatInfo {
+                                    max_ref_mutbl: pat_info.max_ref_mutbl.cap_mutbl_to_not(None),
+                                    ..pat_info
+                                }
+                            } else {
+                                pat_info
+                            };
+                            (expected, r_ty, pat_info)
+                        }
```

Comment:
I am confused by this, does it do what you want? I meant to suggest to move the whole `let pat_info = if r_mutbl == Mutability::Not ...` before the match.

Reply:
Oh, I think I see what you're doing. Is this only to disable diagnostics? Please add a comment explaining that the mutability is already capped but you disable diagnostics to let the borrow-checker catch this error. Or why not catch the error yourself? You're not doing a worse job than the borrow-checker here, and you have more information for suggestions.

Also, I'm pretty sure you can do this in `calc_default_binding_mode` instead, which I would prefer.